### ISSUE TITLE
Anonymous auth breaks ACL plugin

### ISSUE THREAD

**Issue by TehBigA** [Mashape/kong#2547]
### Summary

When an auth plugin assigns a user `anonymous` authorization the ACL plugin does not see this as an assigned consumer and does not pass the request to the upstream application.

### Steps To Reproduce

1. Create anonymous consumer
2. Add `basic-auth` plugin and assign anonymous consumer id to anonymous setting
3. Add `acl` plugin (with or without whitelist/blacklist)
4. Try to hit upstream through kong

_Expected:_ Response from upstream with anonymous auth
_Actual:_ Response from kong
```
{"message":"Cannot identify the consumer, add an authentication plugin to use the ACL plugin"}
```

5. Disable `acl` plugin and retry step 4
_Expected / Actual:_ Response from upstream with anonymous auth

### Additional Details & Logs

- Version: `0.10.1`
- Debug-level startup logs
```
2017/05/22 14:30:13 [verbose] Kong: 0.10.1
2017/05/22 14:30:13 [debug] ngx_lua: 10007
2017/05/22 14:30:13 [debug] nginx: 1011002
2017/05/22 14:30:13 [debug] Lua: LuaJIT 2.1.0-beta2
2017/05/22 14:30:13 [verbose] reading config file at /etc/kong/kong.conf
2017/05/22 14:30:13 [debug] admin_listen = "0.0.0.0:8001"
2017/05/22 14:30:13 [debug] admin_listen_ssl = "0.0.0.0:8444"
2017/05/22 14:30:13 [debug] admin_ssl = true
2017/05/22 14:30:13 [debug] anonymous_reports = true
2017/05/22 14:30:13 [debug] cassandra_consistency = "ONE"
2017/05/22 14:30:13 [debug] cassandra_contact_points = {"127.0.0.1"}
2017/05/22 14:30:13 [debug] cassandra_data_centers = {"dc1:2","dc2:3"}
2017/05/22 14:30:13 [debug] cassandra_keyspace = "kong"
2017/05/22 14:30:13 [debug] cassandra_lb_policy = "RoundRobin"
2017/05/22 14:30:13 [debug] cassandra_port = 9042
2017/05/22 14:30:13 [debug] cassandra_repl_factor = 1
2017/05/22 14:30:13 [debug] cassandra_repl_strategy = "SimpleStrategy"
2017/05/22 14:30:13 [debug] cassandra_ssl = false
2017/05/22 14:30:13 [debug] cassandra_ssl_verify = false
2017/05/22 14:30:13 [debug] cassandra_timeout = 5000
2017/05/22 14:30:13 [debug] cassandra_username = "kong"
2017/05/22 14:30:13 [debug] cluster_listen = "0.0.0.0:7946"
2017/05/22 14:30:13 [debug] cluster_listen_rpc = "127.0.0.1:7373"
2017/05/22 14:30:13 [debug] cluster_profile = "wan"
2017/05/22 14:30:13 [debug] cluster_ttl_on_failure = 3600
2017/05/22 14:30:13 [debug] custom_plugins = {}
2017/05/22 14:30:13 [debug] database = "postgres"
2017/05/22 14:30:13 [debug] dns_hostsfile = "/etc/hosts"
2017/05/22 14:30:13 [debug] dns_resolver = {}
2017/05/22 14:30:13 [debug] log_level = "notice"
2017/05/22 14:30:13 [debug] lua_code_cache = "on"
2017/05/22 14:30:13 [debug] lua_package_cpath = ""
2017/05/22 14:30:13 [debug] lua_package_path = "?/init.lua;./kong/?.lua"
2017/05/22 14:30:13 [debug] lua_socket_pool_size = 30
2017/05/22 14:30:13 [debug] lua_ssl_verify_depth = 1
2017/05/22 14:30:13 [debug] mem_cache_size = "128m"
2017/05/22 14:30:13 [debug] nginx_daemon = "on"
2017/05/22 14:30:13 [debug] nginx_optimizations = true
2017/05/22 14:30:13 [debug] nginx_worker_processes = "auto"
2017/05/22 14:30:13 [debug] pg_database = "kong"
2017/05/22 14:30:13 [debug] pg_host = "127.0.0.1"
2017/05/22 14:30:13 [debug] pg_password = "******"
2017/05/22 14:30:13 [debug] pg_port = 5432
2017/05/22 14:30:13 [debug] pg_ssl = false
2017/05/22 14:30:13 [debug] pg_ssl_verify = false
2017/05/22 14:30:13 [debug] pg_user = "kong"
2017/05/22 14:30:13 [debug] prefix = "/usr/local/kong/"
2017/05/22 14:30:13 [debug] proxy_listen = "0.0.0.0:80"
2017/05/22 14:30:13 [debug] proxy_listen_ssl = "0.0.0.0:8443"
2017/05/22 14:30:13 [debug] serf_path = "serf"
2017/05/22 14:30:13 [debug] ssl = true
2017/05/22 14:30:13 [debug] upstream_keepalive = 60
2017/05/22 14:30:13 [verbose] prefix in use: /usr/local/kong
2017/05/22 14:30:13 [verbose] preparing nginx prefix directory at /usr/local/kong
2017/05/22 14:30:13 [verbose] saving serf identifier to /usr/local/kong/serf/serf.id
2017/05/22 14:30:13 [debug] searching for OpenResty 'resty' executable
2017/05/22 14:30:13 [debug] /usr/local/openresty/bin/resty -V: 'nginx version: openresty/1.11.2.2'
2017/05/22 14:30:13 [debug] found OpenResty 'resty' executable at /usr/local/openresty/bin/resty
2017/05/22 14:30:13 [verbose] saving serf shell script handler to /usr/local/kong/serf/serf_event.sh
2017/05/22 14:30:13 [verbose] SSL enabled, no custom certificate set: using default certificate
2017/05/22 14:30:13 [verbose] default SSL certificate found at /usr/local/kong/ssl/kong-default.crt
2017/05/22 14:30:13 [verbose] Admin SSL enabled, no custom certificate set: using default certificate
2017/05/22 14:30:13 [verbose] admin SSL certificate found at /usr/local/kong/ssl/admin-kong-default.crt
2017/05/22 14:30:13 [warn] ulimit is currently set to "1024". For better performance set it to at least "4096" using "ulimit -n"
2017/05/22 14:30:13 [verbose] running datastore migrations
2017/05/22 14:30:13 [verbose] migrations up to date
2017/05/22 14:30:13 [verbose] serf agent not running, deleting /usr/local/kong/pids/serf.pid
2017/05/22 14:30:13 [debug] searching for 'serf' executable
2017/05/22 14:30:13 [debug] serf version: 'Serf v0.7.0'
2017/05/22 14:30:13 [debug] found 'serf' executable at serf
2017/05/22 14:30:13 [debug] starting serf agent: nohup serf agent -profile 'wan' -bind '0.0.0.0:7946' -log-level 'err' -rpc-addr '127.0.0.1:7373' -event-handler 'member-join,member-leave,member-failed,member-update,member-reap,user:kong=/usr/local/kong/serf/serf_event.sh' -node 'dev_0.0.0.0:7946_470b634076b94e2aa6a0bb7bce7673f7' > /usr/local/kong/logs/serf.log 2>&1 & echo $! > /usr/local/kong/pids/serf.pid
2017/05/22 14:30:13 [verbose] waiting for serf agent to be running
2017/05/22 14:30:14 [debug] sending signal to pid at: /usr/local/kong/pids/serf.pid
2017/05/22 14:30:14 [debug] kill -0 `cat /usr/local/kong/pids/serf.pid` >/dev/null 2>&1
2017/05/22 14:30:14 [verbose] serf agent started
2017/05/22 14:30:14 [verbose] auto-joining serf cluster
2017/05/22 14:30:14 [verbose] no other nodes found in the cluster
2017/05/22 14:30:14 [verbose] registering serf node in datastore
2017/05/22 14:30:14 [verbose] cluster joined and node registered in datastore
2017/05/22 14:30:14 [debug] searching for OpenResty 'nginx' executable
2017/05/22 14:30:14 [debug] /usr/local/openresty/nginx/sbin/nginx -v: 'nginx version: openresty/1.11.2.2'
2017/05/22 14:30:14 [debug] found OpenResty 'nginx' executable at /usr/local/openresty/nginx/sbin/nginx
2017/05/22 14:30:14 [debug] starting nginx: /usr/local/openresty/nginx/sbin/nginx -p /usr/local/kong -c nginx.conf
2017/05/22 14:30:14 [debug] nginx started
2017/05/22 14:30:14 [info] Kong started
```
- Kong error logs:
```
2017/05/22 14:32:26 [notice] 29552#0: using the "epoll" event method
2017/05/22 14:32:26 [notice] 29552#0: openresty/1.11.2.2
2017/05/22 14:32:26 [notice] 29552#0: built by gcc 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.4) 
2017/05/22 14:32:26 [notice] 29552#0: OS: Linux 4.4.0-62-generic
2017/05/22 14:32:26 [notice] 29552#0: getrlimit(RLIMIT_NOFILE): 1024:65536
2017/05/22 14:32:26 [notice] 29564#0: start worker processes
2017/05/22 14:32:26 [notice] 29564#0: start worker process 29567
2017/05/22 14:32:27 [notice] 29567#0: signal 17 (SIGCHLD) received
2017/05/22 14:32:27 [notice] 29567#0: signal 17 (SIGCHLD) received
2017/05/22 14:32:27 [notice] 29567#0: signal 17 (SIGCHLD) received
2017/05/22 14:32:27 [notice] 29567#0: signal 17 (SIGCHLD) received
2017/05/22 14:32:42 [notice] 29567#0: signal 17 (SIGCHLD) received
```
- Kong configuration:

_API:_ Single node.js app that dumps headers as json for testing
_Plugins:_
```
basic-auth (anonymous consumer id set)
jwt (anonymous consumer id set)
acl (whitelist set as 'internal')
```
_Config:_
```
pg_host = 127.0.0.1
pg_port = 5432
pg_user = kong
pg_password = kong
pg_database = kong
```
- Operating System: `Ubuntu 16.04.2 LTS`

**Comment by TehBigA** [comment]
I believe I just discovered the actual issue when trying to fork the ACLs plugin to get around this.

The ACLs plugin looks for `ngx.ctx.authenticated_credential` to exist, but auth plugins such as basic-auth only set `ngx.ctx.authenticated_consumer` for anonymous auth.

**Comment by p0pr0ck5** [comment]
@Tieske any thoughts? I feel like you're familiar with this :)

**Comment by nvx** [comment]
When setting multiple auth plugins (https://github.com/Mashape/kong/issues/590#issuecomment-283627717 describes this) and one auths while the other does not auth, you'd want to make sure you picked the authenticated consumer id rather than the anonymous consumer id set by the other auth plugin. So probably only look at authenticated_consumer if authenticated_credential is not set sounds like the right option?

**Comment by Tieske** [comment]
fixed by #2722, available in 0.11.RC2
